runtime/appruntime/api/server.go

```diff
@@ -0,0 +1,309 @@
+	// forward slash as %2F we don't want the router to treat that as a segment split.
+	//
+	// i.e. `/foo%2Fbar/baz` should be routed to `/:a/*b` as a = "foo/bar", b = "baz"
+	// where as if we use req.URL.Path we would get a = "foo", b = "bar/baz` which is incorrect.
+	path := req.URL.EscapedPath()
+
+	// Switch to the Encore internal router if we are on the Encore internal path
+	const internalPrefix = "/__encore"
+	if strings.HasPrefix(path, internalPrefix+"/") {
+		r = s.encore
+		path = path[len(internalPrefix):] // keep leading slash
+	}
+
+	h, p, _ := r.Lookup(req.Method, path)
+	if h == nil {
+		h, p, _ = r.Lookup(wildcardMethod, path)
+	}
+	if h != nil {
+		// Found an endpoint
+		h(w, req, p)
+		return
+	}
+
+	// Endpoint not found
+	svc, api := "unknown", "Unknown"
+	if idx := strings.IndexByte(ep, '.'); idx != -1 {
+		svc, api = ep[:idx], ep[idx+1:]
+	}
+	metrics.UnknownEndpoint(svc, api)
+	errs.HTTPError(w, errs.B().Code(errs.NotFound).Msg("endpoint not found").Err())
+}
+
+// healthz serves the /__encore/healthz health checking endpoint.
+func (s *Server) healthz(w http.ResponseWriter, _ *http.Request) {
+	w.Header().Set("Content-Type", "application/json")
+
+	bytes, _ := json.Marshal(struct {
+		Code    string `json:"code"`
+		Message string `json:"message"`
+		Details any    `json:"details"`
+	}{
+		Code:    "ok",
+		Message: "Your Encore app is up and running!",
+		Details: struct {
+			AppRevision    string `json:"app_revision"`
+			EncoreCompiler string `json:"encore_compiler"`
+			DeployId       string `json:"deploy_id"`
+		}{
+			AppRevision:    s.cfg.Static.AppCommit.AsRevisionString(),
+			EncoreCompiler: s.cfg.Static.EncoreCompiler,
+			DeployId:       s.cfg.Runtime.DeployID,
+		},
+	})
+	_, _ = w.Write(bytes)
+}
+
+func (s *Server) authPlatformReq(req *http.Request, macSig string) (bool, error) {
+	macBytes, err := base64.RawStdEncoding.DecodeString(macSig)
+	if err != nil {
+		return false, nil
+	}
+
+	// Pull out key ID from hmac prefix
+	const keyIDLen = 4
+	if len(macBytes) < keyIDLen {
+		return false, nil
+	}
+
+	keyID := binary.BigEndian.Uint32(macBytes[:keyIDLen])
+	mac := macBytes[keyIDLen:]
+	for _, k := range s.cfg.Runtime.AuthKeys {
+		if k.KeyID == keyID {
+			return checkAuthKey(k, req, mac), nil
+		}
+	}
+
+	return false, nil
+}
+
+func checkAuthKey(key config.EncoreAuthKey, req *http.Request, gotMac []byte) bool {
+	dateStr := req.Header.Get("Date")
+	if dateStr == "" {
+		return false
+	}
+	date, err := http.ParseTime(dateStr)
+	if err != nil {
+		return false
+	}
+	const threshold = 15 * time.Minute
+	if diff := time.Since(date); diff > threshold || diff < -threshold {
+		return false
+	}
+
+	mac := hmac.New(sha256.New, key.Data)
+	fmt.Fprintf(mac, "%s\x00%s", dateStr, req.URL.Path)
+	expected := mac.Sum(nil)
+	return hmac.Equal(expected, gotMac)
+}
+
+func (s *Server) NewContext(w http.ResponseWriter, req *http.Request, ps PathParams, auth model.AuthInfo) Context {
```

Comment:
Not a fan of that. We've done that in the past but I've always found it quite confusing. I'd rather find a new name for this struct if you're just reacting to it being named "Context"; I don't see any other reason why it would implement `context.Context`